tower-http/src/compression.rs

```diff
@@ -0,0 +1,385 @@
+                    }
+                    Some(Err(e)) => Poll::Ready(Some(Err(Error::Body(e)))),
+                    None => Poll::Ready(None),
+                };
+            }
+        };
+
+        match bytes {
+            Some(Ok(data)) => Poll::Ready(Some(Ok(data.freeze()))),
+            Some(Err(err)) => {
+                let err = inner
+                    .project()
+                    .error
+                    .take()
+                    .map(Error::Body)
+                    .unwrap_or(Error::Compress(err));
+                Poll::Ready(Some(Err(err)))
+            }
+            None => Poll::Ready(None),
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
+        match self.project().inner.project() {
+            BodyInnerProj::Gzip(framed_read) => framed_read
+                .get_pin_mut()
+                .get_pin_mut()
+                .get_pin_mut()
+                .project()
+                .body
+                .poll_trailers(cx),
+            BodyInnerProj::Deflate(framed_read) => framed_read
+                .get_pin_mut()
+                .get_pin_mut()
+                .get_pin_mut()
+                .project()
+                .body
+                .poll_trailers(cx),
+            BodyInnerProj::Brotli(framed_read) => framed_read
+                .get_pin_mut()
+                .get_pin_mut()
+                .get_pin_mut()
+                .project()
+                .body
+                .poll_trailers(cx),
+            BodyInnerProj::Zstd(framed_read) => framed_read
+                .get_pin_mut()
+                .get_pin_mut()
+                .get_pin_mut()
+                .project()
+                .body
+                .poll_trailers(cx),
+            BodyInnerProj::Identity(inner) => inner.poll_trailers(cx),
+        }
+        .map_err(Error::Body)
+    }
+}
+
+#[pin_project(project = BodyInnerProj)]
+#[derive(Debug)]
+enum BodyInner<B, E> {
+    Identity(#[pin] B),
+    Gzip(#[pin] FramedRead<GzipEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
+    Deflate(#[pin] FramedRead<DeflateEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
+    Brotli(#[pin] FramedRead<BrotliEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
+    Zstd(#[pin] FramedRead<ZstdEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
+}
+
+#[pin_project]
+#[derive(Debug)]
+struct Adapter<B, E> {
+    #[pin]
+    body: B,
+    error: Option<E>,
+}
+
+impl<B: Body> Stream for Adapter<B, B::Error> {
+    type Item = io::Result<Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        match ready!(this.body.poll_data(cx)) {
+            Some(Ok(mut data)) => Poll::Ready(Some(Ok(data.copy_to_bytes(data.remaining())))),
```

Comment:
I wonder if there is a way to optimize this if it impls bytes :(

Reply:
I think `<Bytes as Buf>::copy_to_bytes` is already optimized enough. It will become a mere `mem::replace` call if the argument equals `data.remaining()`, without even incrementing the reference count.

<https://github.com/tokio-rs/bytes/blob/master/src/bytes.rs#L551-L559>